### PR TITLE
[WebGL backend] Add max 1D texture dimension flag

### PR DESCRIPTION
**Pull request by Linchenn**
Fix: https://github.com/tensorflow/tfjs/issues/6775
### Problem
Mali GPU could not provide interpolation values with enough precision for long skinny triangles, proved by the [demo](https://github.com/Linchenn/Toolkit/blob/main/GPGPU/gpgpu.js). The varying from the interpolation values will be used to calculate coords (in type) for sampling, finally causing sampling mismatch.

**Standalone code to reproduce the issue**
```js
tf.ENV.set('WEBGL_PACK_DEPTHWISECONV', true)

let w = Array.from({length: 3 * 3 * 816}, () => Math.random())
let x = Array.from({length: 12 * 10 * 816}, () => Math.random())

let inputs = {
    filter: tf.tensor(w, [3, 3, 816, 1]),
    x: tf.tensor(x, [1, 12, 10, 816]),
    strides: 1,
    pad: [[0, 0], [1, 1], [1, 1], [0, 0]],
    dataFormat: "channelsLast",
    dilations: 1,
    activation: 'relu'
};

tf.setBackend('webgl')
tf.ENV.set('WEBGL_MAX_TEXTURE_SIZE', 4096)
let out_4096 = tf.fused.depthwiseConv2d(inputs);

tf.ENV.set('WEBGL_MAX_TEXTURE_SIZE', 2048)
inputs.x = tf.tensor(x, [1, 12, 10, 816])
inputs.filter = tf.tensor(w, [3, 3, 816, 1])
let out_2048 = tf.fused.depthwiseConv2d(inputs);

tf.setBackend('cpu')
inputs.x = tf.tensor(x, [1, 12, 10, 816])
inputs.filter = tf.tensor(w, [3, 3, 816, 1])
let out_reference = tf.fused.depthwiseConv2d(inputs);

const doTensorsDiffer = function(t0, t1) {
	return tf.any(tf.greater(tf.abs(tf.sub(t0, t1)), tf.scalar(1e-2))).dataSync()[0];
}

console.log("Default and 2048 differ? " + doTensorsDiffer(out_4096, out_2048));
console.log("Reference and 2048 differ? " + doTensorsDiffer(out_reference, out_2048));
console.log("Reference and 4096 differ? " + doTensorsDiffer(out_reference, out_4096));
```
The three differs are expected to be 0, while Mali GPU generates 1 for the first and the third.

### Findings
Take the texture with physical size of [3672, 1] for example. As the second dimension increase, the precision becomes better. Actually, when the second dimension is increased to 2, the precision for calculating coords is enough. ([reference data](https://docs.google.com/spreadsheets/d/1uGfdksY1icQJnWEeV9nVX63DHB6bj-xc2IMTjk0sN-Y/edit?usp=sharing): for each table: the second column is the computed index from varying values; the third and forth columns are ground-truth; the fifth column is the difference between indices. You could find all difference between indices are 0.)

### Solution
This PR expose a flag, `WEBGL_MAX_TEXTURE_DIMENSION_FOR_1D_TEXTURE`. For 1D texture (physical height or physical width is 1), if the physical length of any texture edges exceed the threshold, the other edge will be increased to 2.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6808)

### CODE REVIEW

**Comment by Linchenn** [comment]
cc @shanumantesc

**Review by qjia7 — APPROVED** [review]
LGTM, thanks.

Can you also add the `depthwiseConv2d` case described in the bug for webgl?

**Review by Linchenn — COMMENTED** [review]
Done. Thanks!

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6808#-:-NAm_C153QGmUz7CtF6l:boyexb3)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @pyu10055)

**Review by qjia7 — COMMENTED** [review]
Forget to submit the test file?

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6808#-:-NAmb9YW9PM5u3HpKFk9:byueq9h)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @pyu10055)

**Review by pyu10055 — CHANGES_REQUESTED** [review]
Thanks Lin for investigating and fixing this deep hidden bug. I am curious why we take the approach to double the dimension of shape 1, instead of using the squarish function?
I assume the squarish function to produce smaller texture size and potentially reduce the computation if the texture is output.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6808#-:-NAobeVM0KXxSzn9spx6:bpyv5f9)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @Linchenn)

___
*[`tfjs-backend-webgl/src/webgl_util.ts` line 446 at r1](https://reviewable.io/reviews/tensorflow/tfjs/6808#-NAocgGr1ZB3k7CumQW5:-NAocgGsDJ795qR_b_0l:b5r3wzb) ([raw file](https://github.com/tensorflow/tfjs/blob/52dc64443001bbca65252b9311744ad887f310dd/tfjs-backend-webgl/src/webgl_util.ts#L446)):*
> ```XML
>       env().getNumber('WEBGL_MAX_TEXTURE_DIMENSION_FOR_1D_TEXTURE') *
>       (isPacked ? 2 : 1);
>   if (textureShape[0] > maxDimFor1DTex || textureShape[1] > maxDimFor1DTex) {
> ```

It would be good to separate these two conditions to avoid unnecessary computation later.
___
*[`tfjs-backend-webgl/src/webgl_util.ts` line 449 at r1](https://reviewable.io/reviews/tensorflow/tfjs/6808#-NAod10q-FkQ9MbZ50sP:-NAod10q-FkQ9MbZ50sQ:b-k76crw) ([raw file](https://github.com/tensorflow/tfjs/blob/52dc64443001bbca65252b9311744ad887f310dd/tfjs-backend-webgl/src/webgl_util.ts#L449)):*
> ```XML
>     // For 1D texture, if the length exceeds maxDimFor1DTex, the texture will be
>     // upgraded to 2D with a physical shape as [length, 2] or [2, length].
>     textureShape[0] = Math.max(isPacked ? 4 : 2, textureShape[0]);
> ```

should the packed texture be 2?

**Review by Linchenn — COMMENTED** [review]
@qjia7 Since this PR only updates the `getTextureShapeFromLogicalShape` function, I just added some tests for it. 

For the depthwise's mismatch bug here, we do not guarantee this PR could fix it. Instead, this PR expose the flags to users and let users to tune the flag and correct the mismatch.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6808#-:-NApBbrB6pll0BD0KTXi:bthlgid)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @pyu10055)

**Review by pyu10055 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6808)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @Linchenn and @pyu10055)

___
*[`tfjs-backend-webgl/src/webgl_util.ts` line 374 at r2](https://reviewable.io/reviews/tensorflow/tfjs/6808#-NApon02DHH0R7qvamD9:-NApon02DHH0R7qvamDA:ba0xdpj) ([raw file](https://github.com/tensorflow/tfjs/blob/16770325837b02b62374f6eea87ec9dddfa9b71d/tfjs-backend-webgl/src/webgl_util.ts#L374)):*
> ```XML
>       env().getNumber('WEBGL_MAX_SIZE_FOR_NARROW_TEXTURE');
>   if (maxSizeForNarrorTex === Infinity &&
>       env().getBool('WEBGL_AUTO_SQUARIFY_NARROW_TEXTURE_SHAPE_FOR_MALI_GPU')) {
> ```

we need to detect the GPU type if it is MALI, to ensure it only applies to MALI GPU.
or you can change the name to WEBGL_AUTO_SQUARIFY_NARROW_TEXTURE_SHAPE
so it is user's responsibility to detect MALI GPU.
___
*[`tfjs-backend-webgl/src/webgl_util.ts` line 404 at r2](https://reviewable.io/reviews/tensorflow/tfjs/6808#-NApo1et81sxOEECJqRe:-NApo1et81sxOEECJqRf:b-eqnnuw) ([raw file](https://github.com/tensorflow/tfjs/blob/16770325837b02b62374f6eea87ec9dddfa9b71d/tfjs-backend-webgl/src/webgl_util.ts#L404)):*
> ```XML
>       (textureShape: [number, number]) => {
>         return Math.max(...textureShape) > maxSizeForNarrorTex &&
>             Math.min(...textureShape) <= (isPacked ? 2 : 1);
> ```

if the texture is a 0 shaped tensor, we should not change it.
Which means `Math.min(...textureShape) === (isPacked ? 2 : 1)`
___
*[`tfjs-backend-webgl/src/webgl_util.ts` line 420 at r2](https://reviewable.io/reviews/tensorflow/tfjs/6808#-NApoLTIBoZciQW0a4k9:-NApoLTIBoZciQW0a4kA:bh75ehr) ([raw file](https://github.com/tensorflow/tfjs/blob/16770325837b02b62374f6eea87ec9dddfa9b71d/tfjs-backend-webgl/src/webgl_util.ts#L420)):*
> ```XML
>       logShape.length === 2 && logShape[0] <= maxTexSize &&
>       logShape[1] <= maxTexSize &&
>       !isLongNarrowTex(logShape as [number, number])) {
> ```

cache this boolean value, since it is checked in multiple places.

**Review by Linchenn — COMMENTED** [review]
Thank you Ping!

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6808#-:-NAptZGzCYU0BP3uUFnI:b-t1trzh)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @pyu10055)

___
*[`tfjs-backend-webgl/src/webgl_util.ts` line 374 at r2](https://reviewable.io/reviews/tensorflow/tfjs/6808#-NApon02DHH0R7qvamD9:-NApr5Pa43oeEAsvj_2d:bpw4nqf) ([raw file](https://github.com/tensorflow/tfjs/blob/16770325837b02b62374f6eea87ec9dddfa9b71d/tfjs-backend-webgl/src/webgl_util.ts#L374)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

we need to detect the GPU type if it is MALI, to ensure it only applies to MALI GPU.
or you can change the name to WEBGL_AUTO_SQUARIFY_NARROW_TEXTURE_SHAPE
so it is user's responsibility to detect MALI GPU.
</blockquote></details>

Renamed to `WEBGL_AUTO_SQUARIFY_NARROW_TEXTURE_SHAPE`, as we are not sure about when Mali GPU's varying becomes correct. We could add it when we want to maintain this.
___
*[`tfjs-backend-webgl/src/webgl_util.ts` line 404 at r2](https://reviewable.io/reviews/tensorflow/tfjs/6808#-NApo1et81sxOEECJqRe:-NApsDpl5K3oV4O-JNt_:bm5ihck) ([raw file](https://github.com/tensorflow/tfjs/blob/16770325837b02b62374f6eea87ec9dddfa9b71d/tfjs-backend-webgl/src/webgl_util.ts#L404)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

if the texture is a 0 shaped tensor, we should not change it.
Which means `Math.min(...textureShape) === (isPacked ? 2 : 1)`
</blockquote></details>

Since we need to include both `Math.min(...textureShape) === 1` and `Math.min(...textureShape) ===2` for `isPacked === true`, I instead added one more condition here.
___
*[`tfjs-backend-webgl/src/webgl_util.ts` line 420 at r2](https://reviewable.io/reviews/tensorflow/tfjs/6808#-NApoLTIBoZciQW0a4k9:-NApsmWL4XcKvtCdM90h:b-8ufmin) ([raw file](https://github.com/tensorflow/tfjs/blob/16770325837b02b62374f6eea87ec9dddfa9b71d/tfjs-backend-webgl/src/webgl_util.ts#L420)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

cache this boolean value, since it is checked in multiple places.
</blockquote></details>

Since the input for `isLongNarrowTex` is different for each branch, maybe we could not cache it

**Review by Linchenn — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6808)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @pyu10055)

___
*[`tfjs-backend-webgl/src/webgl_util.ts` line 420 at r2](https://reviewable.io/reviews/tensorflow/tfjs/6808#-NApoLTIBoZciQW0a4k9:-NAqnRkt-z2GcSdEsCIA:b-pulve2) ([raw file](https://github.com/tensorflow/tfjs/blob/16770325837b02b62374f6eea87ec9dddfa9b71d/tfjs-backend-webgl/src/webgl_util.ts#L420)):*
<details><summary><i>Previously, Linchenn wrote…</i></summary><blockquote>

Since the input for `isLongNarrowTex` is different for each branch, maybe we could not cache it
</blockquote></details>

Moves checking `isLongNarrowTex` to the end of the function. Then we do not need to add `isLongNarrowTex` to multiple places.

File: tfjs-backend-webgl/src/webgl_util.ts [hunk 1]
```diff
@@ -395,30 +403,42 @@ export function getTextureShapeFromLogicalShape(
   }
 
   let size = util.sizeFromShape(logShape);
-  if (logShape.length <= 1 && size <= maxTexSize) {
-    return [1, size];
+  let textureShape: [number, number];
+  if (logShape.length <= 1 && size <= maxTexSize &&
+      size <= maxSizeForNarrorTex) {
```
Comment (qjia7) [inline tfjs-backend-webgl/src/webgl_util.ts:408]: nit: `size <= maxSizeForNarrorTex` is not needed?

File: tfjs-backend-webgl/src/webgl_util.ts [hunk 2]
```diff
@@ -395,30 +403,42 @@ export function getTextureShapeFromLogicalShape(
   }
 
   let size = util.sizeFromShape(logShape);
-  if (logShape.length <= 1 && size <= maxTexSize) {
-    return [1, size];
+  let textureShape: [number, number];
```
Comment (qjia7) [inline tfjs-backend-webgl/src/webgl_util.ts:406]: nit: `let textureShape = null`. Then the last `else {}` can be removed?

**Review by qjia7 — APPROVED** [review]
LGTM with nits. Thanks.

File: tfjs-backend-webgl/src/webgl_util.ts (hunk 2 above)
Comment (Linchenn) [reply tfjs-backend-webgl/src/webgl_util.ts]: Done.

**Review by pyu10055 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6808)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @pyu10055)

File: tfjs-backend-webgl/src/webgl_util.ts [hunk 3]
```diff
@@ -368,8 +368,16 @@ export function getShapeAs3D(shape: number[]): [number, number, number] {
 export function getTextureShapeFromLogicalShape(
     logShape: number[], isPacked = false): [number, number] {
   let maxTexSize = env().getNumber('WEBGL_MAX_TEXTURE_SIZE');
+  let maxSizeForNarrorTex =
```
Comment (shanumantesc) [inline tfjs-backend-webgl/src/webgl_util.ts:371]:
```suggestion
  let maxSizeForNarrowTex =
```
(same everywhere else)

**Review by shanumantesc — APPROVED** [review]

File: tfjs-backend-webgl/src/webgl_util.ts (hunk 3 above)
Comment (Linchenn) [reply tfjs-backend-webgl/src/webgl_util.ts]: Thanks! Done.

**Pull request merged**